src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -409,10 +409,17 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
             // WSAMsg also contains a single WSABuffer describing a control buffer.
             PinSocketAddressBuffer();
 
-            // Create and pin a WSAMessageBuffer if none already.
+            // Create a WSAMessageBuffer if none exists yet, and ensure the buffer is pinned.
+            Debug.Assert(
+                !_wsaMessageBufferGCHandle.IsAllocated ||
+                (_wsaMessageBufferGCHandle.Target == _wsaMessageBuffer && _wsaMessageBuffer != null));
+            Debug.Assert((_ptrWSAMessageBuffer != IntPtr.Zero) == _wsaMessageBufferGCHandle.IsAllocated);
             if (_wsaMessageBuffer == null)
             {
                 _wsaMessageBuffer = new byte[sizeof(Interop.Winsock.WSAMsg)];
+            }
+            if (_ptrWSAMessageBuffer == IntPtr.Zero)
+            {
```

Comment:
Just for clarity, I think I'd structure the asserts like this -- let me know if I missed something here...

```
if (_wsaMessageBuffer == null)
{
    Debug.Assert(!_wsaMessageBufferGCHandle.IsAllocated);
    Debug.Assert(_ptrWSAMessageBuffer == IntPtr.Zero);

    _wsaMessageBuffer = new byte[sizeof(Interop.Winsock.WSAMsg)];
}
if (_ptrWSAMessageBuffer == IntPtr.Zero)
{
    Debug.Assert(!_wsaMessageBufferGCHandle.IsAllocated);

    _wsaMessageBufferGCHandle = GCHandle.Alloc(_wsaMessageBuffer, GCHandleType.Pinned);
    _ptrWSAMessageBuffer = Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBuffer, 0);
}
else
{
    Debug.Assert(_wsaMessageBufferGCHandle.IsAllocated);
    Debug.Assert(_ptrWSAMessageBuffer == Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBuffer, 0);
}
```

Thoughts?